pallets/automation-time/src/lib.rs

```diff
@@ -955,6 +1090,80 @@ pub mod pallet {
 				.saturating_add(<T as Config>::WeightInfo::run_xcmp_task())
 		}
 
+		/// Executes auto compounding delegation and reschedules task on success
+		pub fn run_auto_compound_delegated_stake_task(
+			delegator: T::AccountId,
+			collator: T::AccountId,
+			account_minimum: BalanceOf<T>,
+			frequency: Seconds,
+			task_id: T::Hash,
+			mut task: Task<T>,
+		) -> (Task<T>, Weight) {
+			match Self::compound_delegator_stake(
+				delegator.clone(),
+				collator.clone(),
+				account_minimum,
+				Self::calculate_execution_fee(&task.action, 1),
+			) {
+				Ok(delegation) =>
+					Self::deposit_event(Event::SuccesfullyAutoCompoundedDelegatorStake {
+						task_id,
+						amount: delegation,
+					}),
+				Err(e) => {
+					Self::deposit_event(Event::AutoCompoundDelegatorStakeFailed {
+						task_id,
+						error_message: Into::<&str>::into(e).as_bytes().to_vec(),
+						error: e,
+					});
+					return (
+						task,
+						<T as Config>::WeightInfo::run_auto_compound_delegated_stake_task(),
+					)
+				},
+			}
+
+			let new_execution_times: Vec<UnixTime> =
+				task.execution_times.iter().map(|when| when.saturating_add(frequency)).collect();
+			let _ = Self::reschedule_existing_task(
+				task_id,
+				task.owner_id.clone(),
+				&task.action,
+				new_execution_times.clone(),
+			)
+			.map(|_| {
+				let new_executions_left: u32 = new_execution_times.len().try_into().unwrap();
+				task.executions_left += new_executions_left;
+				new_execution_times.iter().try_for_each(|t| task.execution_times.try_push(*t))
+			})
+			.map_err(|e| {
+				let err: DispatchErrorWithPostInfo = e.into();
+				Self::deposit_event(Event::AutoCompoundDelegatorStakeFailed {
+					task_id,
+					error_message: Into::<&str>::into(err).as_bytes().to_vec(),
+					error: err,
+				});
+			});
+
+			(task, <T as Config>::WeightInfo::run_auto_compound_delegated_stake_task())
+		}
+
+		fn compound_delegator_stake(
+			delegator: T::AccountId,
+			collator: T::AccountId,
+			account_minimum: BalanceOf<T>,
+			execution_fee: Result<BalanceOf<T>, DispatchError>,
```

Comment:
seems kind of odd to handle this potential error like this

Reply:
Yeah! Fully agreed. But rather than writing a bunch of code to handle that what I think needs to happen is that this function should go back to using `saturating_*` instead of throwing errors: https://github.com/OAK-Foundation/OAK-blockchain/blob/d7b7824f918acda4093c5faad7c3154cba8897b2/pallets/automation-time/src/lib.rs#L1153-L1160

I'm not sure why we changed this to checked+error here. It leads to overly verbose or unnecessary downstream code as you noticed. This method calculates an execution fee. It is reasonable for us to saturate the execution fee on `BalanceOf`. Balance is the same type that's used for total_issuance. We're never overflowing it and even if we did, we should just default to the biggest number.

I'm happy to change the `calculate_execution_fee` method in another pr if my reasoning seems sound.